### ISSUE TITLE
OpenVR on Linux

### ISSUE THREAD

**Issue by ChristophHaag** [highfidelity/hifi#10098]
So I currently have this: https://github.com/ChristophHaag/hifi/commit/f2380d85c551660c0127c8d4ac5b8f3f8c31f7f6
(on top of RELEASE-6227, I'm told it's the latest stable version)

FindOpenVR.cmake is copied from from SteamVR-OSVR and it expects libopenvr_api.so in /usr/lib or anywhere else the linker can find it.

The openvr plugin compiles and loads fine with my changes, but when you try to go into VR mode, interface crashes with
```
[04/01 16:21:52] [FATAL] [default] Cannot make QOpenGLContext current in a different thread
```

This probably needs some changes from someone who knows OpenGL and hifi much better than me.

Note: OpenGL applications are a known issue for the SteamVR implementation on Linux - it has pretty bad performance, but it should work (other OpenGL projects that work are Vivecraft and a Doom 3 fork).

**Comment by humbletim** [comment]
Cool!  I'll test your branch out later today and see if I get the same crash.

Are you able to see controller tracking data from within Interface yet -- or does it crash before things can get that far?

Another approach I've been looking at is using `driver_lighthouse.so` to get tracking data and then doing the rendering the old fashioned way (like by adapting the oculusLegacy plugin).

**Comment by ChristophHaag** [comment]
I'm pretty sure it fails before it can render anything.

If you're going with "client side" rendering, you could get in touch with OpenHMD, they are doing it this way. Full OpenHMD support would be a nice to have thing either way.

**Comment by roothorick** [comment]
The issue is thread-related, obviously. A quirk of OpenGL is that only one thread can interact with a context.

So debugging should start with finding out which thread created the context, which thread is now trying to claim it, and how to move one's context logic into the other.

**Comment by roothorick** [comment]
After a quick googling, you know what? It may be as simple as calling [moveToThread()](http://doc.qt.io/qt-5/qobject.html#moveToThread) before calling makeCurrent(). Depending on where the actual rendering is happening, you may have to call moveToThread() again afterwards at some point.

**Comment by ChristophHaag** [comment]
The OpenGL Vulkan interop extensions are released. Not in the drivers yet, but soon enough. Soon enough OpenGL apps will run with decent performance on SteamVR. Is anyone looking into this issue here?

**Comment by ChristophHaag** [comment]
SteamVR uses these extensions now and gets decent performance in OpenGL applications. Now would be a good time to finish enabling OpenVR support on Linux.

Sadly I'm suffering from frequent GPU hangs with high fidelity (the 2D version), so I'm not going to be able to do much here. https://bugs.freedesktop.org/show_bug.cgi?id=99292

**Comment by ChristophHaag** [comment]
Ported my commit to latest master https://github.com/ChristophHaag/hifi/commit/5eb15334ff3473081b9a8979f14675210c3e949c

Some more windows-only code crept in. Having it enabled on linux, even if it doesn't work yet, would be very helpful to catch this in CI.

Same Qt problem remains.

**Comment by ChristophHaag** [comment]
With threaded submit disabled it worked... at least once.
https://github.com/ChristophHaag/hifi/commit/87cd7a627ce7f4b7558b8aea35fafa3e1a83834b

Kinda unreliable and crashy though.

**Comment by ChristophHaag** [comment]
A little bit of research later: https://stackoverflow.com/a/36804436

tl;dr: Since Qt 4.8 they made makeCurrent() from another thread throw a fatal error but you can enable the old behavior with an attribute:

So threaded submit can be left enabled, with just this one change:
```
diff --git a/interface/src/Application.cpp b/interface/src/Application.cpp
index ea4a3db465..37fb972c4a 100644
--- a/interface/src/Application.cpp
+++ b/interface/src/Application.cpp
@@ -2062,6 +2062,10 @@ Application::~Application() {
 void Application::initializeGL() {
     qCDebug(interfaceapp) << "Created Display Window.";

+    // since Qt 4.8 makeCurrent() only works if you are in the "right" thread
+    // unless the old behavior is enabled like this
+    QCoreApplication::setAttribute(Qt::AA_DontCheckOpenGLContextThreadAffinity);
+
     // initialize glut for shape drawing; Qt apparently initializes it on OS X
     if (_isGLInitialized) {
         return;
```
However, it runs very, very slow and the SteamVR compositor segfaults after a couple of seconds.

So on Linux at least threaded submit is best kept disabled.

**Comment by ChristophHaag** [comment]
Anyone still following this? Looks like my initial pull request to just enable OpenVR on Linux (which would still need several other unrelated patches to prevent crashes) was not received with much interest, so I just keep "maintaining" a VR-enabled version for Linux at the master branch in my fork at https://github.com/ChristophHaag/hifi

Lately it has been segfaulting when enabling VR mode (a ViewFrustum was null) which I found now is caused by commit d5496d68fd84456c2abb96e0c338345e4510b172, so if anyone is interested in looking into this...

**Comment by Casuallynoted** [comment]
I’d also really like to see VR support for Linux be integrated into the Steam version of HiFi. Is there any update on this?

**Comment by source83** [comment]
@ChristophHaag , could you add a small write-up on how and where to add the openvr library so it will be compiled into the interface ?

**Comment by ChristophHaag** [comment]
High Fidelity will download a copy of openvr during running cmake. This seems to happen here: https://github.com/highfidelity/hifi/blob/master/cmake/ports/openvr/portfile.cmake

To use Valve's OpenVR library, hifi comes with an "openvr" plugin. As you can see in this plugin's CMakeLists.txt, it will only be compiled on the "WIN32" platform: https://github.com/highfidelity/hifi/blob/master/plugins/openvr/CMakeLists.txt

In my fork at https://github.com/ChristophHaag/hifi I made it compile on Linux with a bunch of small fixes to make it work. So as far as I know, the high fidelity cmake build system should download and use the openvr library automatically.

**Comment by source83** [comment]
Ok, it is your repository I have been compiling and trying to use, but I have never seen any hint of any VR options anywhere.
Then again, HiFi runs so bad in desktop mode that I'm unsure of if I would actually want to try it in VR on linux with all the lag.
Then again, both on windows and linux (proton) vrchat in desktop mode has always been slow/stuttering for me but smooth in VR, so it might be just fine.
Is there some launch option or some config file that needs to be changed for the OpenVR subsystem to work on your repository @ChristophHaag ?

**Comment by ChristophHaag** [comment]
No, nothing special. If hifi can load its openvr plugin successfully, you should see this menu option
![Screenshot_20190414_204601](https://user-images.githubusercontent.com/781076/56097561-5c004c80-5ef6-11e9-8332-a0234aa27c06.png)

**Comment by source83** [comment]
Trying again from scratch now

**Comment by source83** [comment]
It worked :D
I just spent about 1 hour in hifi with the vive + keyboard and mouse since my controllers were drained over night for some reason :S

**Comment by Codes4Fun** [comment]
Is there any plans to get this into hifi officially?
I've basically duplicated your work not having looked here first, doesn't seem that complicated of a change, at least not with 0.82.1.

**Comment by Codes4Fun** [comment]
I'm also curious how your fork is downloading/using openvr, I had to modify `cmake/ports/hifi-client-deps/CONTROL` file so vcpkg would download it for linux (changing `openvr (windows)` to `openvr (!android)`, and modify `cmake/ports/openvr/portfile.cmake` to be aware of linux and restructure the directory so cmake could find the libraries and headers.

**Comment by ChristophHaag** [comment]
It is indeed not very difficult, the main part is enabling the plugin and making it compile without warnings, which is in my commit "enable openvr plugin on linux and mac os X". The rest is fixing a bunch of random fixes and workarounds on top of that to make sure it actually runs in VR without crashing.

I just tried updating my branch, but it looks like between 0.81 and 0.82 the cmake stuff indeed changed so whatever you did is now necessary.
